# HG changeset patch
# User Torsten <ttl@justmail.de>
# Date 1388870170 -3600
#      Sat Jan 04 22:16:10 2014 +0100
# Branch gui-release
# Node ID 7952fb0435db49b00c4774ab0a33efe936cf0bfe
# Parent  55b8dc2ea0222ab6e201b6fbb8155003478bc461
improve title bar of dock widgets

* libgui/src/module.mk: new icon files for darker dock widget's title bars
* libgui/src/resource.qrc: new icon files for darker dock widget's title bars
* octave-dock-widget.cc (constructor): class vars for close action and icons;
  (set_title): transparent background of title text;
  (make_window, make_widget): new dock icon depends on chosen color;
  (handle_settings): title bar with gradient, chose icons depending on color
* octave-dock-widget.h: new class variables

diff --git a/libgui/src/module.mk b/libgui/src/module.mk
--- a/libgui/src/module.mk
+++ b/libgui/src/module.mk
@@ -66,16 +66,19 @@ octave_gui_ICONS = \
   src/icons/stop.png \
   src/icons/terminal.png \
   src/icons/undo.png \
   src/icons/up.png \
   src/icons/warning.png \
   src/icons/widget-close.png \
   src/icons/widget-dock.png \
   src/icons/widget-undock.png \
+  src/icons/widget-close-light.png \
+  src/icons/widget-dock-light.png \
+  src/icons/widget-undock-light.png \
   src/icons/zoom-in.png \
   src/icons/zoom-out.png
 
 octave_gui_MOC =
 
 if AMCOND_HAVE_QSCINTILLA
 octave_gui_MOC += \
   src/m-editor/moc-file-editor-interface.cc \
diff --git a/libgui/src/octave-dock-widget.cc b/libgui/src/octave-dock-widget.cc
--- a/libgui/src/octave-dock-widget.cc
+++ b/libgui/src/octave-dock-widget.cc
@@ -60,26 +60,28 @@ octave_dock_widget::octave_dock_widget (
   _dock_action-> setToolTip (tr ("Undock widget"));
   connect (_dock_action, SIGNAL (triggered (bool)),
            this, SLOT (change_floating (bool)));
   _dock_button = new QToolButton (this);
   _dock_button->setDefaultAction (_dock_action);
   _dock_button->setFocusPolicy (Qt::NoFocus);
   _dock_button->setIconSize (QSize (12,12));
 
-  QAction *close_action = new QAction
+  _close_action = new QAction
                    (QIcon (":/actions/icons/widget-close.png"), "", this );
-  close_action-> setToolTip (tr ("Hide widget"));
-  connect (close_action, SIGNAL (triggered (bool)),
+  _close_action-> setToolTip (tr ("Hide widget"));
+  connect (_close_action, SIGNAL (triggered (bool)),
            this, SLOT (change_visibility (bool)));
   _close_button = new QToolButton (this);
-  _close_button->setDefaultAction (close_action);
+  _close_button->setDefaultAction (_close_action);
   _close_button->setFocusPolicy (Qt::NoFocus);
   _close_button->setIconSize (QSize (12,12));
 
+  _icon_color = "";
+
   QHBoxLayout *h_layout = new QHBoxLayout ();
   h_layout->addStretch (100);
   h_layout->addWidget (_dock_button);
   h_layout->addWidget (_close_button);
   h_layout->setSpacing (0);
   h_layout->setContentsMargins (5,2,2,2);
 
   _title_widget = new QWidget ();
@@ -141,16 +143,17 @@ octave_dock_widget::connect_visibility_c
 // set the title in the dockwidgets title bar
 void
 octave_dock_widget::set_title (const QString& title)
 {
 #if defined (Q_OS_WIN32)
   QHBoxLayout* h_layout =
     static_cast<QHBoxLayout *> (titleBarWidget ()->layout ());
   QLabel *label = new QLabel (title);
+  label->setStyleSheet ("background: transparent;");
   h_layout->insertWidget (0,label);
 #endif
   setWindowTitle (title);
 }
 
 // make the widget floating
 void
 octave_dock_widget::make_window ()
@@ -165,30 +168,30 @@ octave_dock_widget::make_window ()
   // FIXME: dockWidgetArea always returns 2
   settings->setValue ("DockWidgets/" + objectName () + "_dock_area",
                       _parent->dockWidgetArea (this));
   settings->setValue ("DockWidgets/" + objectName (), saveGeometry ());
   settings->sync ();
 
   // remove parent and adjust the (un)dock icon
   setParent (0, Qt::Window);
-  _dock_action->setIcon (QIcon (":/actions/icons/widget-dock.png"));
+  _dock_action->setIcon (QIcon (":/actions/icons/widget-dock"+_icon_color+".png"));
   _dock_action->setToolTip (tr ("Dock widget"));
 
   // restore the last geometry( when floating
   setGeometry (settings->value ("DockWidgets/" + objectName ()
                        + "_floating_geometry",QRect(50,100,480,480)).toRect ());
 
 #else
 
   // non windows: Just set the appripriate window flag
   setWindowFlags (Qt::Window);
 
   QString css = styleSheet ();
-  css.replace ("widget-undock.png","widget-dock.png");
+  css.replace ("widget-undock","widget-dock");
   setStyleSheet (css);
 
 #endif
 
   _floating = true;
 }
 
 // dock the widget
@@ -218,26 +221,26 @@ octave_dock_widget::make_widget (bool do
       //        and its child widget
       restoreGeometry (settings->value
              ("DockWidgets/" + objectName ()).toByteArray ());
     }
   else  // only reparent, no docking
     setParent (_parent);
 
   // adjust the (un)dock icon
-  _dock_action->setIcon (QIcon (":/actions/icons/widget-undock.png"));
+  _dock_action->setIcon (QIcon (":/actions/icons/widget-undock"+_icon_color+".png"));
   _dock_action->setToolTip (tr ("Undock widget"));
 
 #else
 
   // non windows: just say we are a docked widget again
   setWindowFlags (Qt::Widget);
 
   QString css = styleSheet ();
-  css.replace ("widget-dock.png","widget-undock.png");
+  css.replace ("widget-dock","widget-undock");
   setStyleSheet (css);
 
 #endif
 
   _floating = false;
 }
 
 // slot for (un)dock action
@@ -271,75 +274,85 @@ octave_dock_widget::focusWidget ()
 }
 
 void
 octave_dock_widget::handle_settings (const QSettings *settings)
 {
   QString css;
   QString css_button;
   QString dock_icon;
+
   if (_floating)
-    dock_icon = "widget-dock.png";
+    dock_icon = "widget-dock";
   else
-    dock_icon = "widget-undock.png";
+    dock_icon = "widget-undock";
 
   if (settings->value ("DockWidgets/widget_title_custom_style",false).toBool ())
     {
 
       QColor default_var = QColor (0,0,0);
       QColor fg_color = settings->value ("Dockwidgets/title_fg_color",
                                          default_var).value<QColor> ();
 
       default_var = QColor (255,255,255);
       QColor bg_color = settings->value ("Dockwidgets/title_bg_color",
                                          default_var).value<QColor> ();
 
-      QString bg_icon = QString ("transparent");
-      if (bg_color.lightness () < 128)
-        bg_icon = fg_color.name ();
+      int r, g, b;
+      QColor bg_color_light = bg_color.lighter ();
+
+      bg_color.getRgb (&r, &g, &b);
+      if (r+g+b < 400)
+          _icon_color = "-light";
+      else
+        _icon_color = "";
+
+      QString background =
+          QString ("background: qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1,"
+                   "            stop: 0 %1, stop: 0.75 %2, stop: 0.9 %2, stop: 1.0 %1);").
+                   arg (bg_color_light.name ()).
+                   arg (bg_color.name ());
 
 #if defined (Q_OS_WIN32)
-      css = QString ("background: %1; color: %2 ;").
-                     arg (bg_color.name ()).
-                     arg (fg_color.name ());
-      css_button = QString ("background: %3; border: 0px;").arg (bg_icon);
+      css = background + QString (" color: %1 ;").arg (fg_color.name ());
 #else
-      css = QString ("QDockWidget::title { background: %1;"
+      css = QString ("QDockWidget::title { " + background +
                      "                     text-align: center left;"
                      "                     padding: 0px 0px 0px 4px;}\n"
-                     "QDockWidget { color: %2 ; "
-                     "  titlebar-close-icon: url(:/actions/icons/widget-close.png);"
-                     "  titlebar-normal-icon: url(:/actions/icons/"+dock_icon+"); }"
+                     "QDockWidget { color: %1 ; "
+                     "  titlebar-close-icon: url(:/actions/icons/widget-close%2.png);"
+                     "  titlebar-normal-icon: url(:/actions/icons/"+dock_icon+"%2); }"
                      "QDockWidget::close-button,"
-                     "QDockWidget::float-button { background: %3; border: 0px;}"
+                     "QDockWidget::float-button { border: 0px;}"
                      ).
-                     arg (bg_color.name ()).
                      arg (fg_color.name ()).
-                     arg (bg_icon);
+                     arg (_icon_color);
 #endif
     }
   else
     {
 #if defined (Q_OS_WIN32)
       css = QString ("");
-      css_button = QString ("background: transparent; border: 0px;");
 #else
       css = QString ("QDockWidget::title { text-align: center left;"
                      "                     padding: 0px 0px 0px 4px;}"
                      "QDockWidget {"
                      "  titlebar-close-icon: url(:/actions/icons/widget-close.png);"
                      "  titlebar-normal-icon: url(:/actions/icons/"+dock_icon+"); }"
                      "QDockWidget::close-button,"
                      "QDockWidget::float-button { border: 0px; }"
                     );
 #endif
     }
 
 #if defined (Q_OS_WIN32)
   _title_widget->setStyleSheet (css);
+  css_button = QString ("background: transparent; border: 0px;");
   _dock_button->setStyleSheet (css_button);
   _close_button->setStyleSheet (css_button);
+  _dock_action->setIcon (QIcon (":/actions/icons/"+dock_icon+_icon_color+".png"));
+  _close_action->setIcon (QIcon (":/actions/icons/widget-close"+_icon_color+".png"));
 #else
   setStyleSheet (css);
 #endif
 
   notice_settings (settings);  // call individual handler
 }
diff --git a/libgui/src/octave-dock-widget.h b/libgui/src/octave-dock-widget.h
--- a/libgui/src/octave-dock-widget.h
+++ b/libgui/src/octave-dock-widget.h
@@ -105,20 +105,22 @@ protected slots:
 private slots:
 
   void change_floating (bool);
   void change_visibility (bool);
 
 private:
 
   QMainWindow *_parent;  // store the parent since we are reparenting to 0
-  QAction *_dock_action;
   bool _floating;
+  QString _icon_color;
 
 #if defined (Q_OS_WIN32)
   QWidget *_title_widget;
   QToolButton *_dock_button;
   QToolButton *_close_button;
+  QAction *_dock_action;
+  QAction *_close_action;
 #endif
 
 };
 
 #endif
diff --git a/libgui/src/resource.qrc b/libgui/src/resource.qrc
--- a/libgui/src/resource.qrc
+++ b/libgui/src/resource.qrc
@@ -62,10 +62,13 @@
         <file>icons/graphic_logo_HistoryDockWidget.png</file>
         <file>icons/graphic_logo_WorkspaceView.png</file>
         <file>icons/graphic_logo_DocumentationDockWidget.png</file>
         <file>icons/graphic_logo_ReleaseWidget.png</file>
         <file>icons/warning.png</file>
         <file>icons/widget-close.png</file>
         <file>icons/widget-dock.png</file>
         <file>icons/widget-undock.png</file>
+        <file>icons/widget-close-light.png</file>
+        <file>icons/widget-dock-light.png</file>
+        <file>icons/widget-undock-light.png</file>
     </qresource>
 </RCC>
